### ISSUE TITLE
Crash issue in certain missions in builds 20200315 and later

### ISSUE THREAD

**Issue by Kerbrer** [scp-fs2open/fs2open.github.com#2316]
When attempting to commit to the mission Apocalypse in FreeSpace Blue (and only FreeSpace Blue) with builds 20200315 and higher, the game just abruptly quits to desktop 100% of the time. The log simply states that a crash was detected. The dump mentions a memory exception at 0xC05 and that the build didn't have access to that address.

Apocalypse is a red-alert mission.

[EDIT, added from comment below:] This doesn't happen in the vanilla version of Apocalypse, only the FSBlue version. There's also a red-alert mission in Mantle where this crash seems to happen on committing, according to MatthTheGeek, and in Aftermath 2, the game crashes during the opening cutscene while panning past a ship's fighterbay very early on.

The files linked are from 20200401, but the logs and the dump are identical to the 20200315 files.

Log:
https://fsnebula.org/log/5e8b2a21cb0d3321b0c9863a
Crash Dump:
https://cdn.discordapp.com/attachments/255067124836335616/696710518882435142/fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG_20200406_130836.mdmp

**Comment by asarium** [comment]
So, it appears that a ship does not have a primary bank weapon and so the game crashes upon trying to use the weapon index `-1`:
```
>	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!get_max_ammo_count_for_primary_bank(int ship_class, int bank, int ammo_type) Line 16068	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!ship_process_post(object * obj, float frametime) Line 8859	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!obj_move_all_post(object * objp, float frametime) Line 1243	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!obj_move_all(float) Line 1498	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!game_simulation_frame() Line 3526	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!game_frame(bool paused) Line 3925	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!game_do_state(int state) Line 5951	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!gameseq_process_events() Line 383	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!game_main(int argc, char * * argv) Line 6427	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!SDL_main(int argc, char * * argv) Line 7466	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!main_getcmdline(...) Line 177	C
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!invoke_main() Line 99	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!__scrt_common_main_seh() Line 253	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!__scrt_common_main() Line 296	C++
 	fs2_open_20_1_0_20200401_8cb9d31_x64_SSE2-FASTDBG.exe!WinMainCRTStartup() Line 17	C++
```

@Goober5000 This appears to be related to you vector changes although that did not touch this code so I guess that it just kind of worked before for some weird reason.

**Comment by Kerbrer** [comment]
Forgot to mention that this doesn't happen in the vanilla version of Apocalypse, only the FSBlue version. There's also a red-alert mission in Mantle where this crash seems to happen on committing, according to MatthTheGeek, and in Aftermath 2, the game crashes during the opening cutscene while panning past a ship's fighterbay very early on.

**Comment by Goober5000** [comment]
@asarium Yeah, the difference is that SCP_vector actually does bounds checking.  Previously, -1 references would apparently silently work.  I wonder why various memory-checking analyses never flagged this.  A -1 reference is also the cause of the weapon cheat crash.

**Comment by tomimaki** [comment]
PR fixes issue for Apocalypse in FSBlue, but opening cutscene in Aftermath 2 still crashes.

```
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!std::vector<weapon_info,std::allocator<weapon_info>>::operator[](const unsigned __int64 _Pos) Wiersz 1500
	w C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.25.28610\include\vector(1500)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!all_turret_weapons_have_flags(ship_weapon * swp, Weapon::Info_Flags flags) Wiersz 282
	w C:\fs2open.github.com\code\ai\aiturret.cpp(282)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!ai_fire_from_turret(ship * shipp, ship_subsys * ss, int parent_objnum) Wiersz 2227
	w C:\fs2open.github.com\code\ai\aiturret.cpp(2227)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!process_subobjects(int objnum) Wiersz 11182
	w C:\fs2open.github.com\code\ai\aicode.cpp(11182)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!ai_frame(int objnum) Wiersz 14073
	w C:\fs2open.github.com\code\ai\aicode.cpp(14073)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!ai_process(object * obj, int ai_index, float frametime) Wiersz 14178
	w C:\fs2open.github.com\code\ai\aicode.cpp(14178)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!ship_process_post(object * obj, float frametime) Wiersz 9002
	w C:\fs2open.github.com\code\ship\ship.cpp(9002)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!obj_move_all_post(object * objp, float frametime) Wiersz 1243
	w C:\fs2open.github.com\code\object\object.cpp(1243)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!obj_move_all(float frametime) Wiersz 1498
	w C:\fs2open.github.com\code\object\object.cpp(1498)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!game_simulation_frame() Wiersz 3526
	w C:\fs2open.github.com\freespace2\freespace.cpp(3526)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!game_frame(bool paused) Wiersz 3925
	w C:\fs2open.github.com\freespace2\freespace.cpp(3925)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!game_do_frame() Wiersz 4286
	w C:\fs2open.github.com\freespace2\freespace.cpp(4286)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!game_do_state(int state) Wiersz 5951
	w C:\fs2open.github.com\freespace2\freespace.cpp(5951)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!gameseq_process_events() Wiersz 381
	w C:\fs2open.github.com\code\gamesequence\gamesequence.cpp(381)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!game_main(int argc, char * * argv) Wiersz 6426
	w C:\fs2open.github.com\freespace2\freespace.cpp(6426)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!SDL_main(int argc, char * * argv) Wiersz 7466
	w C:\fs2open.github.com\freespace2\freespace.cpp(7466)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!main_getcmdline(...) Wiersz 177
	w c:\projects\sdl\src\main\windows\sdl_windows_main.c(177)
```

**Comment by tomimaki** [comment]
#2318 fixes Aftermath 2 cutscene.

But there is another crash connected to #2294. Mission One Future in Blue Planet WiH. Crash occurs during switching weapons.
```
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!std::vector<weapon_info,std::allocator<weapon_info>>::operator[](const unsigned __int64 _Pos) Wiersz 1500
	w C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.25.28610\include\vector(1500)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!ConditionedHook::ConditionsValid(int action, object * objp, int more_data) Wiersz 477
	w C:\fs2open.github.com\code\scripting\scripting.cpp(477)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!script_state::RunCondition(int action, object * objp, int more_data) Wiersz 864
	w C:\fs2open.github.com\code\scripting\scripting.cpp(864)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!ship_select_next_primary(object * objp, int direction) Wiersz 12377
	w C:\fs2open.github.com\code\ship\ship.cpp(12377)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!button_function_critical(int n, net_player * p) Wiersz 1819
	w C:\fs2open.github.com\code\io\keycontrol.cpp(1819)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!button_function(int n) Wiersz 2356
	w C:\fs2open.github.com\code\io\keycontrol.cpp(2356)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!button_info_do(button_info * bi) Wiersz 2882
	w C:\fs2open.github.com\code\io\keycontrol.cpp(2882)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!game_process_keys() Wiersz 1747
	w C:\fs2open.github.com\code\io\keycontrol.cpp(1747)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!game_frame(bool paused) Wiersz 3900
	w C:\fs2open.github.com\freespace2\freespace.cpp(3900)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!game_do_frame() Wiersz 4286
	w C:\fs2open.github.com\freespace2\freespace.cpp(4286)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!game_do_state(int state) Wiersz 5951
	w C:\fs2open.github.com\freespace2\freespace.cpp(5951)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!gameseq_process_events() Wiersz 381
	w C:\fs2open.github.com\code\gamesequence\gamesequence.cpp(381)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!game_main(int argc, char * * argv) Wiersz 6426
	w C:\fs2open.github.com\freespace2\freespace.cpp(6426)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!SDL_main(int argc, char * * argv) Wiersz 7466
	w C:\fs2open.github.com\freespace2\freespace.cpp(7466)
fs2_open_20_1_0_x64_AVX2-DEBUG.exe!main_getcmdline(...) Wiersz 177
	w c:\projects\sdl\src\main\windows\sdl_windows_main.c(177)
```

**Comment by tomimaki** [comment]
Mission One Future in BP2 fixed by #2319.

**Comment by Goober5000** [comment]
All reported crashes should now be fixed.  Wait for the next nightly build, and then let me know if there are any more.
